### DISCUSSION TITLE
Implementation of 'hdr-compute-blacklevel' that adapts to the Darkest Pixel in Scene and Spreads the Tone-Mapping Curve reaching Enhanced Contrasts

### DISCUSSION THREAD

**Discussion by stonix015** [haasn/libplacebo#303] [Ideas]
**EDIT:
See my second and last comment for a solution that is useable, but not "clean and nice" regarding the code implementation** 

Hey @haasn ,

I am thinking of a feature that measures the lowest spot in the scene and scales the scenes, so that this spot appears as a 0 nits // 0 through RGB values...
...Like the opposide of the hdr-compute-peak feature:
Setting the **target-contrast in mpv** raises the black level of the output format to the specs of the desired display by **upscaling the scene from the src black-level to the display black-level**.
In the actual version of libplacebo, the src black-level is always assumed to be 0 meaning that a constant upscaling is taking place for every movie and every scene in it.
But what if a scene **only has pixels that are above 0**?
In this scene, the black-level scaling of mpv would not only be insufficient, but also harming the image quality by a visible greyish look.
A _black-level-detection_ could work as the **opposide pendant to the hdr-peak-detection**:
In this case, the scene would be less upscaled, adapting its specific darkest pixel to the desired display black-level.

**My personal Case** _(Could be indirectly solved by using the feature)_
Surprisingly but disappointingly lots of movies have black levels in the greyish area. Even in darkest scenes, a full black level will not be reached. Instead of, the scene looks washed out. 
Using some gamma tricks to raise the details in the dark (e.g. to make movies watchable in projector setups) makes the problem worse: Now the "grey" level of the movie looks bright grey instead of dark grey.
But those settings are the only values with which some other dark movies can be enjoyed. 

The result would be a contrastier image, which quality would be uniform and indepent from the image quality of the movie.
Very much greetings,
Lukas

**Comment by stonix015** [comment]
Hey,

I tried my best the last two days understanding the code.

My try was to search for the peak-detect feature and have a look how it works:

From my understanding, the function is `measure_peak` in `colorspace.c:1027` .
The part of the function ...
``` 
for (int k = 0; k < SLICES; k++)
    next += data->frame_hist[k][i];
if (next < target_pixel) {
    sum = next;
    continue;
}
```  
... looks for me like pixels with a raising brightness are counted, until there are so many pixels counted as the peak-percentile values suggests (`target_pixel`).

From that perspective i think, that `data->frame_hist[k][i]` stores the number/count of pixels sorted as a 2 dimensional array regarding a rising PQ brightness aquivalent, where `data->frame_hist[0][0]` means "0 nits" and `data->frame_hist[SLICES][HIST_BINS]` means "10'000nits".

From that perspective, i thought of a function like:
``` 
static float measure_min(const struct peak_buf_data *data)
{

    //Src min Black level
    for (int k = 0; k < SLICES; k++)
    {
        if (data->frame_hist[k][0] >= 1) {continue;}
    }
    const float ratio = k / SLICES;

    const float pq_low  = (float) HIST_PQ(-64) / PQ_MAX;
    const float pq_high       = (float) HIST_PQ(-63) / PQ_MAX;
    const float min_black = PL_MIX(pq_low, pq_high, ratio);

    return new_black;
}
```  
where just the lowest bin (-64) is refered with its slices from 0 to SLICES.
I empirically determined, that this should reach for brightness values from 0 nits to around 0,1nits. That might do the trick for most movies with slightly greyish dark scenes where a total black_level of 0 is NOT reached. (Thats what i want to implement this feature for).


Now i have some troubles: I am missing the idea how to transfer the `new_black` value to the `spline function` // `tone_mapping.c`.
There are also some plot holes in my thoughts, e.g. why can HIST_PQ (x) be counted from `x=-64` to `x = 64`, while `data->frame_hist[0][0]` is counted from zero -> Shouldn't the `i` be shifted of those value of -64 at `colorspace.c:1063` :
``` 
// PQ luminance associated with count_low/high respectively
const float pq_low  = (float) HIST_PQ(i)     / PQ_MAX;
float pq_high       = (float) HIST_PQ(i + 1) / PQ_MAX;
```

I think that's all i can do at this place.
My head is crushed now. :/

Very many greetings!
Lukas

**Comment by stonix015** [comment]
Hey @haasn ... if you had interest in the feature, we can discuss this via discord. I made it to work now with a few workarounds:

In my comment above, i already stated an idea how to implement this thing. The only thing was missing to me was knowledge how to bring the value from `colorspace.c` to the `tone_mapping.c`.
Now i worked around that! I used the `input_avg` variable instead and declined the adaptive avg-feature of the spline filter.
(This is not worse, as i already worked with the linear tone-mapping before, there are multiple reasons why the spline tone-mapping does not sufficient on my very dark projector-setup)

**Creating the new black level value (shaders/colorspace.c)**
First, i had to change the histogram enum-function to create a histogram from 0nits to HDR completely:
```
enum {
    SLICES = 12,
    PQ_BITS     = 14,
    PQ_MAX      = (1 << PQ_BITS) - 1,
    HIST_BITS   = 7,
    HIST_BIAS   = 0, //Change here!
    // In the version before, the histogram was splitted into <100nits and >100nits. 
    // Everything under 100nits wasnt practically useable.
    HIST_BINS   = (1 << HIST_BITS) - HIST_BIAS,
#define HIST_PQ(bin) (((bin) + HIST_BIAS) << (PQ_BITS - HIST_BITS))
};
```

secondly, i had to create a `measure_black` function, which works similar to the `measure_peak`:
```
static float measure_black(const struct peak_buf_data *data, float percentile)
{
    unsigned next = 0;
    float slice_ratio = 0;
    for (int i = 0; i < HIST_BINS; i++) {
        for (int k = 0; k < SLICES; k++){
            next += data->frame_hist[k][i];
            if (next < 10000) //Counterpart to the hdr-peak-percentile, the first 10'000 pixels are ignored.
                slice_ratio = k; // will be updated until 10'000 pixels are reached. Yes! The if function will be entered every time until that.
            //most movies have slight noise, therefore this is senseful to clip them.
        }
        if (next < 1000)
            continue;
        const float ratio = slice_ratio / SLICES;
        const float pq_low  = (float) HIST_PQ(i) / PQ_MAX+0.000001f; // somehow, a total zero would throw the peak detection out.
        const float pq_high = (float) HIST_PQ(i+1) / PQ_MAX;
        return PL_MIX(pq_low, pq_high, ratio);
    }
}
```

Then, i had to change the avg-function in the `update_peak_buf`
```
float avg_pq, max_pq;
if (frame_wg_active) {
    avg_pq = measure_black(&data, params->percentile);
    max_pq = measure_peak(&data, params->percentile);
} else {
    // Solid black frame
    avg_pq = max_pq = PL_COLOR_HDR_BLACK;
}
```





**Using the new black-level value (tone_mapping.c)**
After that, i just had to use the value in the `tone_mappin.c` :

First, i want to limit the feature to the specified output_min level.
If the scene black-level appears to be above the display-spec, it should still appear as "grey":
```
const float new_input_min = PL_MIN(params->output_min, params->input_avg);
```

It should be said firstly, that a use in the linear-function (`rescale_in` and `rescale_out`) didnt have any effect. So i decided to transfer the `spline` function into a kind of linear function.

Therefore, i had to modify the `st2094_pick_knee` function: 
```
[...]
// New Line here:
const float new_input_min = PL_MIN(params->output_min, params->input_avg);
// Modified line here:
const float src_min = pl_hdr_rescale(params->input_scaling,  PL_HDR_PQ, new_input_min);
const float src_max = pl_hdr_rescale(params->input_scaling,  PL_HDR_PQ, params->input_max);
// Modified line here:
const float src_avg = pl_hdr_rescale(params->input_scaling,  PL_HDR_PQ, new_input_min); 
const float dst_min = pl_hdr_rescale(params->output_scaling, PL_HDR_PQ, params->output_min);
const float dst_max = pl_hdr_rescale(params->output_scaling, PL_HDR_PQ, params->output_max);
[...]
```

And then, i had to modify the `spline` function:
```
[...]
// New Line here:
const float new_input_min = PL_MIN(params->output_min, params->input_avg);
// Modified Line here:
float slope = (dst_pivot - params->output_min) / (src_pivot - new_input_min);
float ratio = params->input_max / params->output_max - 1.0f;
ratio = fclampf(params->constants.slope_tuning * ratio,
                params->constants.slope_offset,
                1.0f + params->constants.slope_offset);
slope = powf(slope, (1.0f - params->constants.spline_contrast) * ratio);
// Normalize everything the pivot to make the math easier
// Modified Line here:
const float in_min = new_input_min - src_pivot;
const float in_max = params->input_max - src_pivot;
const float out_min = params->output_min - dst_pivot;
const float out_max = params->output_max - dst_pivot;
[...]
```





**Thats all.**
The spline function works similar to the linear function now, and the minimum black level in picture will be detected.
From practical use, i can say that LOTS OF scenes appear more contrasty than before.

**Comment by stonix015** [comment]
Some examples of the feature (it might help to show the screenshots in fullscreen and Alt+Tabbing them):
Prey with the feature enabled.
![prey_with](https://github.com/user-attachments/assets/19065a3a-bb4c-49f8-bce0-138facfe52cc)
Prey with the old version of libplacebo
![prey_without](https://github.com/user-attachments/assets/d2bf2bb3-5483-420c-a4dd-483984b2355a)

And also as a movie which already has 0-black levels:
Bodies Bodies Bodies with the feature enabled
![bodies_with](https://github.com/user-attachments/assets/a8410ad8-fdd2-48aa-8654-670b54d074c7)
And with the feature disabled:
![bodies_without](https://github.com/user-attachments/assets/5dde38e1-d59d-412b-8d89-7153c89da2cc)


You can clearly see the difference in Prey, which is very washed out. You can also see that this feature adapts to the scenes black level, as it doesnt do anything in Bodies Bodies Bodies.

**Reply by haasn** [reply]
Hey; sorry for not replying yet as I have been rather busy. I do find this feature worth consideration; and I can help implement it properly. (The underlying logic is straightforward as is just an inverse of the histogram peak)

I will get back to you on this

**Reply by stonix015** [reply]
No stress, if you want we can discord and talk about that (in german).

would be very cool!

:)
Lukas

**Comment by DisplayTalk** [comment]
Hey @stonix015 , can you link the modified libplacebo source files? I tried putting in those changes and building mpv with my libplacebo fork and no dice(it fails to build, maybe I messed something up). I want to try this because there's a lot of old anime with non zero black level. It would be so nice for it to be taken care of automatically.

**Reply by stonix015** [reply]
Hey @DisplayTalk,

To make it short: I have developed this thing so it runs on my streamer-pc. But it is nothing official and it is nothing that has been pushed officially. It also might not work with actual versions of ffmpeg, mpv, glslang, etc.

Are you a linux-user? I could donate you the modified package via discord (if you have a look at my profile, you can add me).
I have compiled it with arch linux, that might be the easiest way. I can also create an updated version, means i could use recent sources for that. In this moment i just deal with the "old" version and explicitely do NOT update my streamer pc, so i do not crash anything.

Are you a windows user? Than i do not know how to compile this.

Greetings,
Lukas

**Reply by stonix015** [reply]
By the way: Since my last postings here, there were a few modifications. I had noticed a few flickerings and some scenes, in which the feature kind of failed. I found out some big mistakes (some mistake be like: Why the f* hell did that even work?!).

I also added a feature that adapts to the rate of level changes to adapt for scene changes. Aim of the feature is a direct adaption of the new black level in a new scene, but to smooth the black level changes during a running scene. That really helped. Because before you only had the choice between a direct adaption (pointed out to flicker badly in some scenes) or a smoothed adaption (pointed out to "a fading effect" in new scenes that have a significantly other black level)
This feature is very advanced, as it uses the black-level, the mean brightness value and the peak brightness value of the scene and controls every value for bigger changes. In a new scene, i assume that all of those three values significantly changes. if only one value changes, this might indicate that during one scene just the sun pops out behind a rooftop or something like that.

**Reply by DisplayTalk** [reply]
I tried following the discord link both on my phone and pc and it doesn't work. I am a linux user but I really need it to work on windows since that's where my rife setup is right now.

**Reply by DisplayTalk** [reply]
@stonix015 Just fork libplacebo on your profile and add your changes to the repository, then I can fork your repo. That will let me build mpv for windows with your changes to libplacebo.

**Reply by stonix015** [reply]
Hey! I've taken note of that and will proceed. Please bear with me for a moment, as I have a shitload of work to tackle on my end, haha. However, I am willed to share my version of this filter more publicly with those who might be interested.

Creating a fork seems like a great option for me to do this without undermining Niclas. It's important for me to emphasize that libplacebo will remain as is, with Niclas credited as the original creator. I want to clarify that my contribution is solely the filter, which is still imperfect and somewhat raw in terms of code.

I apologize for not being well-versed in GitHub; I'm just starting to learn these things now! As an electrical engineer, we don’t typically deal much with GitHub...

Greetings
Lukas (Luuuks, Luke, whatever you like)

**Reply by DisplayTalk** [reply]
No problem, take your time, I just recently learned it as well. Much easier to build a fork of mpv for windows with github actions than other methods. I needed to do that to take advantage of the recent fix-desat change by @haasn for sdr to hdr/hdr to hdr inverse tonemapping on my tv. Your idea here would be the cherry on top, and it's something I've wished for for a good long time after having to manually adjust black level on so many different movies.

**Reply by DisplayTalk** [reply]
https://discord.gg/hU426vtU

@stonix015 Can you join my discord so we can get connected on there? I can help with the fork, I have lots of time.

**Reply by kwerner72** [reply]
Hi,
I tried this feature also and was really pleased with the improved image quality.
The contrast in general really was better than in without the patch. Though in some bright scenes, the image was getting too bright.

Then I wanted to combine this patch, with a dynamic peak percentile, proposed in the mpv repo: [need help with lua script (dynamic hdr-peak-percentile)](https://github.com/mpv-player/mpv/discussions/16290)
Without this patch, the values for "video-params/avg-pq-y" were reasonable. Using this patch, these values were way lower, even though the image itself looked quite similar.

I also tried your "HDR-compute-peak" branch, @stonix015 
With this, the values was a bit higher. But for my taste, the image was too dark.

Please dont get me wrong. Dont want to complain about a patch not working with some other experimental feature.
Just wanted to make sure, there is nothing broken in this patch.
And also wanted to ask, if there are some other news regarding the blacklevel feature.

Greetings,
Kai

**Reply by stonix015** [reply]
> Hi, I tried this feature also and was really pleased with the improved image quality. The contrast in general really was better than in without the patch. Though in some bright scenes, the image was getting too bright.
> 
> Then I wanted to combine this patch, with a dynamic peak percentile, proposed in the mpv repo: [need help with lua script (dynamic hdr-peak-percentile)](https://github.com/mpv-player/mpv/discussions/16290) Without this patch, the values for "video-params/avg-pq-y" were reasonable. Using this patch, these values were way lower, even though the image itself looked quite similar.
> 
> I also tried your "HDR-compute-peak" branch, @stonix015 With this, the values was a bit higher. But for my taste, the image was too dark.
> 
> Please dont get me wrong. Dont want to complain about a patch not working with some other experimental feature. Just wanted to make sure, there is nothing broken in this patch. And also wanted to ask, if there are some other news regarding the blacklevel feature.
> 
> Greetings, Kai


Hey kwerner72,

I am really sorry for that, but combining with other patches i am totally out. I dont know what this patch does and how things mash up.

The problem is, that i am still waiting for an official libplacebo branch, where haasn helps implementing the feature. 
It shouldnt be much efford. The idea is already completed, he just needs to merge this and to fix 2 variables that i have used the wrong way. 
To clarify: I do only have understood about 2 percent of the code, or maybe 3. I have reused the peak-detect feature, because it already gives an idea, how the black-detect feature could look like. And then i used some C#-skills i learned at the university in the first two semesters. 
In other words: My feature is really isolated in some lines of code that i modified. I dont know what all the other lines an functions in libplacebo do.......... 

Greetings,
stonix

**Reply by kwerner72** [reply]
Hi stonix,

thanks four your reply!!
I also tried to understand the parts of the code, necessary to improve black level.

As you mentioned, you started with using measure_peak and adapted it for measuring the true black level. Then you took this value and stored it in avg_pq, so it can be used later in spline functions.

I took your measure_black function and added an new min_pq variable, to not override avg_pq. The new min_pq is then used to set hdr min value (colorspace.c:799), which was PL_COLOR_HDR_BLACK before.

Another necessary change was to lower the black level the higher the avg_pq is:
_min_pq = min_pq - (min_pq * avg_pq);_

With these changes the resulting image had (almost) the black levels of your version. But also same bright areas as the regular libplacebo version.

With the patch from above
![dune2_2_bl](https://github.com/user-attachments/assets/c247db2e-ba88-420b-833e-25e85d2e6781)

my version
![dune2_2_custom](https://github.com/user-attachments/assets/f95aa1e8-e55f-4d1f-8c6a-eb5b7af3beb5)

master branch
![dune2_2_master](https://github.com/user-attachments/assets/4665f7db-ce0c-4e04-90a1-9872cf510f3d)


With the patch from above
![val_1_bl](https://github.com/user-attachments/assets/c4dfec88-aaa8-4901-8d13-b6e03528fe55)

my version
![val_1_custom](https://github.com/user-attachments/assets/a6b7d2ad-57d6-4056-81c9-5bc218970f86)

master branch
![val_1_master](https://github.com/user-attachments/assets/5fc7bdcf-a798-46da-9a6c-b6d0624aaf44)

To make the lower boundary configurable, there is now also a new parameter "black_percentile". Dont like hardcoded values ;o)

If you want, you can check these changes here:
https://github.com/kwerner72/libplacebo/tree/dyn-black


Also I'm quite curious on how you plan to implement this feature.
If I can help implementing or testing, feel free to contact me :o)

Greetings,
Kai

**Reply by DisplayTalk** [reply]
This is good stuff, although it's flickery. Looks like you removed the peak decay rate feature but didn't replace it with anything. Also for some reason still requires using hdr-peak-percentage even though you added a separate black-percentage

**Reply by stonix015** [reply]
> This is good stuff, although it's flickery. Looks like you removed the peak decay rate feature but didn't replace it with anything. Also for some reason still requires using hdr-peak-percentage even though you added a separate black-percentage

Hey @DisplayTalk :

To sum up what happened now with @kwerner72:

We put our heads together and fully wrapped up the feature. The missing puzzle piece in the version I’d been using until now was figuring out how to pass the detected BlackLevel value (`min_pq`) from `<shaders/colorspace.c>` over to `<tone_mapping.c>`.

Since the spline tone mapping already passed along the average scene brightness (`scene_avg`), I repurposed that variable to sneak in the black level instead. That way, I could get the BlackLevel info through to where I needed it — even though it technically wasn’t meant for that.

As a side effect, the real `scene_avg` value was no longer available in `<tone_mapping.c>`, which ended up turning the spline mapping into something that behaved more like linear tone mapping. (Along with some hacky workaround code in `<tone_mapping.c>` to make it work.)

So, in my version, BlackLevel detection was working fine!
But I couldn’t take advantage of what the spline was actually meant to do — using `scene_avg` to dynamically scale content to match the display’s target peak brightness and compress highlights accordingly.

kwerner72 helped me figure out exactly how and where in the code the pq-values get passed from `<shaders/colorspace.c>` to `<tone_mapping.c>`:
They are passed from `<shaders/colorspace.c>` to `<colorspace.c>` (directly in src dir) and then processed to `<tone_mapping.c>`!
The key was that in the `<colorspace.c>` was a function in which `min_luma`, `avg_luma` and `max_luma` were defined by zero, `avg_pq_y`, and `max_pq_y`. Now the key was to set `min_luma = min_pq_y` and add the `min_pq_y` to the header files and in the output-function of the `<shaders/colorspace.c>`.
Changes in the `<tone_mapping.c>` are now not longer requested, it automatically deals with the `min_luma` value that is now correct. One would say: I correctly advanced haasns code that already was ready to go before.
One would say the `min_luma` variable was like a template....

**Reply by stonix015** [reply]
@DisplayTalk 

You can see the code in the last pullrequest officially here in the repository. As i finally tried to merge this (dont know where haasn is now) to the final code. That code, that can simply be loaded with `pacman -Sy libplacebo mpv`.

What you expected, that the peak-decay-rate was removed, is not the state here.
I only reprogrammed it with a more sophisticated scene-chance-detection.
(A hdr-peak scene-change will be detected if `avg` and `max` both underlie a bigger change rate than defined in the config file., A hdr-black scene-change will be detected if `avg` and `min` both underlie a bigger change rate than defined).

what you could do now is to merge the pullrequest into your libplacebo fork and compile it.
But also changes in mpv sources are needed, as new variables were added to the config-file.

I will stop helping people personally here, because its too much work for me. Instead of, i will try to merge those code, which now works really really great, into the official repos.
If this, what i did here, will be merged into libplacebo, i can merge the config-code into mpv.

**Reply by DisplayTalk** [reply]
That one does work pretty well(still some flicker from sudden changes) with  
hdr-peak-decay-rate=20
 hdr-scene-threshold-low=5.5
 hdr-scene-threshold-high=10.0
 
 and libplacebo opts:
 black_maxadvance=8,black_percentile=3
 
 I am using inverse tonemapping and spline. When using no inverse tonemapping bt.2446a seems best(but insanely crushed with these settings on inverse tonemapping).
 
This needs a setting added in mpv to enable/disable the black detection, and it needs the peak-percentile value disconnected from it. Currently it still only works with peak-percentile enabled to some value less than 100 even though the black-percentile setting was added. That part doesn't make sense to me.

**Reply by DisplayTalk** [reply]
I can also see now that it's still not working on "clip" tonemap. Should be able to use the black detection while only clipping for higher values of the pq curve if desired by using the clip setting. Edit: worse than that, it actually raises the black level if you try to use it with clip

**Comment by stonix015** [comment]
Hey @kwerner72,
 
I think your git name + Kai sounds a bit like a german to me?
Maybe we can talk via discord (in german). Your advances sound very interesting and sophisticated for me.
 
But whats most interesting for me is that you added the min_pq variable, because i didnt know how to. (Transferring the variable from colorspace.c to tone_mappings.c). Therefore i kind of deactivated the true spline tone-mapping, which is very interesting when watching HDR content with projectors in home theatres (very low brightness here).
 
I' thank you very much! Maybe we can team up :D
I hope haasn doesnt interpret this as peer pressure......
 
Greetings,
stonix

**Comment by stonix015** [comment]
@kwerner72 
Update (No nerd leaves to holidays without taking his notebook with him)
I have read your code. It seems to be really straight forwarded. So the variables are transferred from shaders/colorspace.c (where they will be generated) via src/colorspace.c (one directory above) to params->XYZ and then read by tone_mapping.c.
 
So the consequent question would be: How could ive been so silly not to understand this before..... ?! :D 
But yes! it gives me new possibilities. Ill check that when i am home. Maybe then i can recompile a new version of libplacebo, where the spline filter correctly works.
 
One thing:Can you explain to me what you do with the line? :

`min_pq = min_pq - (avg*min_pq)`

I do not understand where the error in calculations would be without this line ......... 
 
Greetings and i am really really thankful for your help!,
stonix015

**Reply by kwerner72** [reply]
Hi stonix,

great you like the changes :o)

The line for lowering the black level was necessary, because otherwise very bright images would have been to dark.
You can see it here...

master branch:
![valerian_master](https://github.com/user-attachments/assets/64ef1547-4352-4e9b-9ef4-b3ba3540f9cf)

With black level detection
![valerian_staticblack](https://github.com/user-attachments/assets/903e3f4f-d147-442b-b4f9-9b53729a6ef6)

With dynamic black level detection
![valerian_dynblack](https://github.com/user-attachments/assets/3f6a94a1-5a94-471c-80d4-33fd39a7b4e9)

Currently I'm testing some non-linear reduction functions.
My current favorite is:
min_pq = PL_MAX(min_pq - (min_pq * PL_MIN((avg_pq * avg_pq * 4), 1.0f)), PL_COLOR_HDR_BLACK);

So every image above 0.5 avg_pq will have default black levels, whereas dark images will have approximately the detected black level.

Sure, we  can talk about that on discord.
Du solltest schon eine Anfrage von einem "kalten Tier" haben ;o)

**Comment by stonix015** [comment]
Hey,

now i understand your point!

it is really senseful for me as i am a projector user and blacks will be killed by inroom response on bright images. 

my approach was just to lower the min-pq value by 1/100 * avg, where 1/100 is the approx room "chess-field" Contrast.

**Comment by stonix015** [comment]
The discord invite isnt shown to me on mobile (android). I have to try on my computer. We are standing in the Stau on the Autobahn.

**Reply by stonix015** [reply]
@kwerner72 i searched all discord options on computer now, but it seems he is hiding your request on my profile ..... maybe you can join @DisplayTalk s channel https://discord.gg/bYn5gZEa and we can meet there? Just as an idea.

Or you can try texting me instead of just "adding" me....
EDIT:
now i have allowed all of this in discord, since some month it is even not allowed in discord for strangers to text ..... the options standard setting is deactivated for allowing it.

**Comment by SuRGeoNix** [comment]
Hey guys, I’ve recently reworked [Flyleaf's renderer](https://github.com/SuRGeoNix/Flyleaf/blob/master/FlyleafLib/MediaFramework/MediaRenderer/Renderer.VP.FL.HDR.cs) to use spline tone mapping with dynamic per-frame luminance analysis (min / avg / peak), including temporal smoothing and scene-change handling. 

The work was partly inspired by libplacebo’s peak detection and the black-level detection work discussed here / in the related PR.

I thought you might be interested to see it in action, and any feedback is more than welcome.

Here’s a quick fair comparison with mpv using the same target parameters on my SDR display: 270-nit peak and 0.5-nit minimum:

Flyleaf (Auto configured to monitor's min/peak | Can be configured also from Flyleaf.Config.json -> TargetMinNits/TargetMaxNits)
<img width="3840" height="2160" alt="flyleaf" src="https://github.com/user-attachments/assets/19374cf9-dc77-4539-9cfc-ebee5f6722db" />

MPV (with --vo=gpu-next --hdr-reference-white=270 --target-contrast=540)
<img width="1920" height="1080" alt="mpv" src="https://github.com/user-attachments/assets/4db18e76-21e8-4f64-bd79-ec76fe6fcedb" />

**Reply by stonix015** [reply]
hey dude!
This is unbelievable for me that my work has just inspired someone to adapt it and use it in his environment.

Unfortunately everything went to sleep ever since, and i use it on my streamer desk, which is still an arch from a year ago without any updates. Doing updates would do a big mess yet, since all of the packages clash due to their versions. Everything has to be rebuild and recoded once again then. I didnt have any time for that for now, but i cant see the point in it either.

An implementation in the official libplacebo would be great. I could do regular updates then.... 'caugh.

But, too many projects, from pinball machines to old cars, hifi, PhD thesis, and so on ..... 

Greetings! Its really a joy for me that someone invests in this feature after all of the time!
stonix

**Reply by kasper93** [reply]
For the record, I also think that it may be useful in some cases to have a black level detection. But there are some tricky issues to solve there. Also this is on far end of my todo list, because there are more prominent fixes and improvements that I'd like to land first, to have a solid base, before adding another layers and point of tm curve adjustment on top. For example this series https://code.videolan.org/videolan/libplacebo/-/merge_requests/875 and some more changes I have locally will make a big difference too.

**Reply by SuRGeoNix** [reply]
> For the record, I also think that it may be useful in some cases to have a black level detection. But there are some tricky issues to solve there. Also this is on far end of my todo list, because there are more prominent fixes and improvements that I'd like to land first, to have a solid base, before adding another layers and point of tm curve adjustment on top. For example this series https://code.videolan.org/videolan/libplacebo/-/merge_requests/875 and some more changes I have locally will make a big difference too.

@kasper93 Here what I've learned during the new HDR implementation that might help:

- Don't let static HDR (such as Content Light / Mastering Display) metadata drive your tone mapper
- Measure the actual frames and use metadata as supplementary information
- Know the real display black/peak range (Using 0 nits for my monitor's ~0.5-nit black level makes a huge difference)
- Compute a lightweight asynchronous (1-3 frames older data are fine as a performance tradeoff) luminance histogram
- Derive robust min/avg/peak scene statistics, feed those into a proper luminance-aware tone mapper such as spline, perform black/peak adaptation, and stabilize the resulting parameters temporally with explicit scene-change detection
- Use target min/peak aware delinearizer instead of pure gamma 2.2 (I replaced it with finite-black inverse BT.1886)

If everything works as it should, you don't need tone-mapping hacks or weird parameters. Just give the spline the right source and target data and let it do its job.

And I do agree, accurate black-level detection is the cherry on top.
